### PR TITLE
PT-4118 refresh traits job

### PR DESCRIPTION
**Pull request by alexbukata**

### CODE REVIEW

File: enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/repository/TraitRepository.kt [hunk 1]
```diff
@@ -42,6 +48,11 @@ class TraitRepository(
         return template.find<Trait>(Query(criteria)).collectList().awaitFirst()
     }
 
+    suspend fun deleteAllByCollection(collectionId: EnrichmentCollectionId) {
+        val query = Query(where(Trait::collectionId).`is`(collectionId.toString()))
```
Comment (dmitry-saraykin) [inline enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/repository/TraitRepository.kt:52]: isEqualTo. 
Are you sure you need to do toString here?

File: enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/service/TraitService.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import org.slf4j.LoggerFactory
+import org.springframework.stereotype.Component
+
+@Component
+class TraitService(
+    private val itemRepository: ItemRepository,
+    private val traitRepository: TraitRepository
+) {
+    suspend fun recalculateTraits(collectionId: EnrichmentCollectionId) {
+        traitRepository.deleteAllByCollection(collectionId)
```
Comment (yakischik) [inline enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/service/TraitService.kt:15]: Is there any chance some traits might be created somewhere in listener while we recalculationg? In such case insert all might not work

File: worker/src/main/kotlin/com/rarible/protocol/union/worker/task/RefreshTraitsTaskHandler.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                val shouldRun = AtomicReference(true)
+                while (shouldRun.get()) {
+                    val collections =
+                        collectionRepository.findAll(fromIdExcluded = lastId.get(), limit = BATCH_SIZE).toList()
+                    collections.forEach { collection ->
+                        try {
+                            logger.info("Processing collection ${collection.id}")
+                            traitService.recalculateTraits(collection.id)
+                        } catch (ex: Exception) {
+                            logger.warn("Error processing collection for updating traits", ex)
```
Comment (yakischik) [inline worker/src/main/kotlin/com/rarible/protocol/union/worker/task/RefreshTraitsTaskHandler.kt:41]: Better to log failed collection id

File: enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/repository/ItemRepository.kt [hunk 4]
```diff
@@ -190,6 +198,42 @@ class ItemRepository(
         ShortItem::class.java
     ).asFlow()
 
+    suspend fun getTraitsByCollection(collectionId: EnrichmentCollectionId): List<Trait> {
+        val traitsCount = ConcurrentHashMap<UnionMetaAttribute, Int>()
+        val listedTraitsCount = ConcurrentHashMap<UnionMetaAttribute, Int>()
+        val query = Query(where(ShortItem::collectionId).`is`(collectionId.toString()))
```
Comment (dmitry-saraykin) [inline enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/repository/ItemRepository.kt:204]: this probably also will not work. As far as I remember we have it only for custom collections

File: enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/repository/TraitRepository.kt (hunk 1 above)
Comment (alexbukata) [reply enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/repository/TraitRepository.kt:52]: done

File: enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/service/TraitService.kt (hunk 2 above)
Comment (alexbukata) [reply enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/service/TraitService.kt:15]: done

File: worker/src/test/kotlin/com/rarible/protocol/union/worker/task/RefreshTraitsTaskHandlerIt.kt [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                )
+        )
+
+        refreshTraitsTaskHandler.runLongTask(null, "").toList()
+
+        val afterRefresh = mongo.findAll(Trait::class.java).asFlow().toList()
+
+        assertThat(afterRefresh).containsExactlyInAnyOrder(
+            Trait(
+                collectionId = collection.id,
```
[the remark this answers is not in the document]
Comment (alexbukata) [reply worker/src/test/kotlin/com/rarible/protocol/union/worker/task/RefreshTraitsTaskHandlerIt.kt:95]: I don't see it
Actually I use all fields here, so don't see why I should use randomTrait() here

File: enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/repository/ItemRepository.kt (hunk 4 above)
Comment (dmitry-saraykin) [reply enrichment/src/main/kotlin/com/rarible/protocol/union/enrichment/repository/ItemRepository.kt:204]: example from prod union db: ETHEREUM:0x0000000005756b5a03e751bd0280e3a55bc05b6e:10000 - no collectionId
also there is no index for collectionId

File: listener/src/test/kotlin/com/rarible/protocol/union/listener/service/EnrichmentItemEventServiceIt.kt [hunk 6]
```diff
@@ -115,6 +117,7 @@ class EnrichmentItemEventServiceIt : AbstractIntegrationTest() {
 
         val unionItem = EthItemConverter.convert(ethItem, itemId.blockchain)
         val shortItem = ShortItemConverter.convert(unionItem).copy(
+            collectionId = AddressFactory.create().toString(),
```
Comment (yakischik) [inline listener/src/test/kotlin/com/rarible/protocol/union/listener/service/EnrichmentItemEventServiceIt.kt:120]: randomAddress()

**Review by yakischik — APPROVED** [review]

**Review by sanufriev — APPROVED** [review]

**Pull request merged**